lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2382 @@
+        batch_size, n_obs_steps = batch["observation.state"].shape[:2]
+        assert n_obs_steps == self.config.n_obs_steps
+
+        # Extract image feature (first combine batch and sequence dims).
+        img_features = self.rgb_encoder(einops.rearrange(batch["observation.image"], "b n ... -> (b n) ..."))
+        # Separate batch and sequence dims.
+        img_features = einops.rearrange(img_features, "(b n) ... -> b n ...", b=batch_size)
+        
+        # image observation feature, state feature, and action query token are grouped together with the same timestpe to form a group, which is listed in order to be entered into GPT sequentially.
+        observation_feature = torch.cat([
+                torch.unsqueeze(self.obs_projector(img_features), dim=2), 
+                torch.unsqueeze(self.state_projector(batch["observation.state"]), dim=2), 
+                self._action_token.repeat(batch_size, n_obs_steps, 1, 1)
+            ], dim=-2).view(batch_size, -1, self.config.gpt_input_dim)
+        if img_features.shape[1] != n_obs_steps:
+            raise NotImplementedError
+        len_additional_action_token = self.config.n_action_pred_token-1
+        action_token = self._action_token.repeat(batch_size, len_additional_action_token, 1)
+
+        # add additional action query tokens for predicting future action chunks
+        observation_feature = torch.cat([observation_feature, action_token], dim=1)
+
+        
+        # get action features (pass through GPT)
+        features = self.policy(observation_feature)
+        historical_act_pred_index = np.arange(0, n_obs_steps) * (self.config.gpt_num_obs_mode+1) + self.config.gpt_num_obs_mode
+
+        # only extract the output tokens at the position of action query
+        features = torch.cat([
+            features[:, historical_act_pred_index],
+            features[:, -len_additional_action_token:]
+        ], dim=1)
+        # pass through action head
+        pred_action = self.action_head(
+            features,
+        )
+        # if rollout, VQ-BeT don't calculate loss
+        if rollout:
+            return pred_action["predicted_action"][:, n_obs_steps-1, :].reshape(batch_size, self.config.n_action_pred_chunk, -1)
+        # else, it calculate overall loss (bin prediction loss, and offset loss)
+        else:
+            action = batch["action"]
+            n, total_w, act_dim = action.shape
+            act_w = self.config.n_action_pred_chunk
+            num_token = total_w + 1 - act_w
+            output_shape = (n, num_token, act_w, act_dim)
+            output = torch.empty(output_shape).to(action.device)
+            for i in range(num_token):
+                output[:, i, :, :] = action[:, i : i + act_w, :]
+            action = output
+
+            loss = self.action_head.loss_fn(
+                pred_action,
+                action,
+                reduction="mean",
+            )
```

Comment:
Do you mind if we simplify here?

First we need this in the `__init__`:

```python
num_tokens = self.config.n_action_pred_token + self.config.n_action_pred_chunk - 1
self.register_buffer(
                "select_target_actions_indices",
                torch.row_stack(
                    [torch.arange(i, i + self.config.n_action_pred_chunk) for i in range(num_tokens)]
                ),
            )
```

This saves us from recreating the tensor each forward pass. We actually have a performance related task in to do this with ACT https://github.com/orgs/huggingface/projects/46?pane=issue&itemId=64971003.

Then here we can cut down a lot (might need some verification on your part):

```suggestion
        else:
            output = batch["action"][:, self.select_target_actions_indices]
            loss = self.action_head.loss_fn(pred_action, output, reduction="mean")
            return pred_action, loss
```

Reply:
Btw it's also okay to create the indices on the spot if you prefer that to using a buffer (it might be more readable). But they should be created with `device=batch['action'].device` in order to create them directly on GPU rather than creating them on CPU (because then we need a device transfer).

Reply:
FYI here's a thread where I'm trying to understand the performance implications: https://discord.com/channels/1216765309076115607/1237737604611571722/1246144412329709578. But I would still say my suggestion above mostly stands because I think it makes the code easier to read.

Reply:
Thank you for your suggestion! I changed following your code.